socketio_middleware/src/rooms.rs

```diff
@@ -33,35 +35,114 @@ impl ChannelPair {
 }
 
 pub fn join_channel_to_room(room_id: &str, channel_pair: ChannelPair) {
+    //TODO: why use remove, it is low performance.
     let mut connected_sockets = match ROOMS.remove(room_id) {
         Some(val) => val,
         None => Vec::new(),
     };
 
-    connected_sockets.push(channel_pair);
+    //check if socketid exist
+    let mut exist: bool = false;
+    for socket in &connected_sockets {
+        if socket.sid() == channel_pair.sid() {
+            debug!("ROOMS: socketid {} doesn't join into room {}, this socketid already exist in the room.", channel_pair.sid(), room_id);
+            exist = true;
+            break;
+        }
+    }
 
+    if false == exist {
```

Comment:
Let's change this to `if !exists {`. Comparing directly to a boolean is more verbose than we want.

Reply:
changed